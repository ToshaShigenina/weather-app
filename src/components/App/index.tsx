import React, { FC } from "react";
import { Sidebar } from "../Sidebar";
import { TabsWeather } from "../TabsWeather";
import { DetailsWeather } from "../DetailsWeather";
import { Page } from "../UI/Page";
import { ThemeContextProvider } from "../UI/Theme";
import { RootStoreProvider } from '../../store';

export const App: FC = () => {
  return (
    <RootStoreProvider>
      <ThemeContextProvider>
        <Page sidebar={ <Sidebar /> }>
          <TabsWeather />
          <DetailsWeather />
        </Page>
      </ThemeContextProvider>
    </RootStoreProvider>
  );
};
