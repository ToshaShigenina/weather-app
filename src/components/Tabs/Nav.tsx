import React from "react";
import cn from "classnames";

export interface NavProps extends Omit<React.OlHTMLAttributes<HTMLUListElement>, 'children' | 'onClick'> {
	active: number,
	tabsList: string[],
	onClick?: (index: number) => void
}

const Nav: React.FC<NavProps> = ({ active = 0, tabsList = [], className, onClick, ...otherProps }) => {
	const onClickHandler = (index: number) => {
		if (onClick) {
			onClick(index);
		}
	};

	return (
		<>
			{
				Boolean(tabsList.length) &&
				<ul { ...otherProps } className={ cn(className, 'tabs__nav') }>
					{
						tabsList.map((item, i) => {
							return (
								<li className="tabs__nav-item" key={ 'tabs-nav-item-' + i }>
									<button
										className={ cn('tabs__nav-btn', active === i && 'active') }
										onClick={ () => onClickHandler(i) }
									>
										{ item }
									</button>
								</li>
							)
						})
					}
				</ul>
			}
		</>
	);
};

export default Nav;