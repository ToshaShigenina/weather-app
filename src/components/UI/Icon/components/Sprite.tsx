import React, { FC, SVGAttributes, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { ReactComponent as SpriteSvg } from '../svg/sprite.svg'

import '../css/sprite.css';

const portalNode = document.createElement('div');
portalNode.setAttribute('id', 'sprite');
portalNode.setAttribute('data-testid', 'sprite');
portalNode.hidden = true;

export const Sprite: FC<SVGAttributes<SVGElement>> = (props) => {
	useEffect(() => {
		document.body.appendChild(portalNode);
		return () => portalNode.remove();
	});
	return createPortal(<SpriteSvg { ...props } />, portalNode)
};
