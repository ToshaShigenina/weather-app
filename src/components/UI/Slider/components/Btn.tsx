import React, { ButtonHTMLAttributes, FC } from 'react';
import cn from 'classnames';
import { ReactComponent as SliderArrow } from '../svg/slider-arrow.svg'

type SliderBtnDirectionType = 'prev' | 'next';
interface SliderBtnProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    direction: SliderBtnDirectionType
}

export const Btn: FC<SliderBtnProps> = ({ direction = '', className, ...otherProps }) => {
    return (
        <button
            className={ cn(className, 'slider__btn', direction && `slider__btn-${direction}`) }
            { ...otherProps }
        >
            <SliderArrow />
        </button>
    );
};
