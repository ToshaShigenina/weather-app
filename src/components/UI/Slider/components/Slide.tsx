import React, { FC, HTMLAttributes, useContext, useEffect, useState } from 'react';
import cn from 'classnames';
import { SliderContext, type SliderContextType } from '../context';

export const Slide: FC<HTMLAttributes<HTMLDivElement>> = ({ children, className = '', style = {}, ...otherProps }) => {
    const slideKey = Symbol();
    const [slideIndex, setSlideIndex] = useState(0);
    const sliderCtx = useContext<SliderContextType | null>(SliderContext);
    const width = sliderCtx && sliderCtx.width > 0
        ? sliderCtx.width + 'px'
        : '100%';
    const marginLeft = sliderCtx?.margin || 0;

    useEffect(() => {
        if (sliderCtx && sliderCtx.addSliderKey && !sliderCtx.sliderKeys.includes(slideKey)) {
            setSlideIndex(sliderCtx.sliderKeys.length);
            sliderCtx.addSliderKey(slideKey);
        }
    }, []);
    return (
        <div
            className={ cn(className, 'slide') }
            style={ {
                ...style,
                width,
                marginLeft: slideIndex ? marginLeft : 0,
            } }
            { ...otherProps }
        >
            { children }
        </div>
    );
};
