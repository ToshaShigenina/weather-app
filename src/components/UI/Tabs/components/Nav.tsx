import React, { FC, OlHTMLAttributes } from "react";
import cn from "classnames";

export interface NavProps extends Omit<OlHTMLAttributes<HTMLUListElement>, 'children' | 'onClick'> {
	active?: number,
	items?: string[],
	onClick?: (index: number) => void
}

export const Nav: FC<NavProps> = ({ active = 0, items = [], onClick, ...othrProps }) => {
	const onClickHandler = (index: number) => {
		if (onClick) {
			onClick(index);
		}
	};

	return (
		<>
			{
				Boolean(items.length) &&
				<ul className={ 'tabs-nav' } { ...othrProps }>
					{
						items.map((item, i) => {
							return (
								<li className={ cn('tabs-nav__item', active === i && 'active') } key={ 'tabs-nav-item-' + i }>
									<button
										className="tabs-nav__btn"
										onClick={ () => onClickHandler(i) }
									>
										{ item }
									</button>
								</li>
							)
						})
					}
				</ul>
			}
		</>
	);
};
