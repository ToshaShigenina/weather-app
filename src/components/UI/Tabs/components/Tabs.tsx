import React, { FC, HTMLAttributes, ReactNode, useState } from "react";
import cn from "classnames";
import { Nav } from './Nav';
import '../style.css';

export type TabItem = {
	label: string
	children: ReactNode
};
export interface TabsProps extends Omit<HTMLAttributes<HTMLDivElement>, 'children'> {
	items: TabItem[]
};
type LabelsFunc = (items: TabItem[]) => string[];

const getLabels: LabelsFunc = (items) => {
	return items.map(item => item.label);
};
export const Tabs: FC<TabsProps> = ({ items = [] }) => {
	const [active, setActive] = useState(0);

	const changeActiveTab = (index: number) => {
		setActive(index);
	};

	return (
		<>
			<Nav active={ active } items={ getLabels(items) } onClick={ changeActiveTab } />
			<div className="tabs-content">
				{ items[active].children }
			</div>
		</>
	);
};
