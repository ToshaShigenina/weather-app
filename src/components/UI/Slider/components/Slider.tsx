import React, { FC, HTMLAttributes, useRef, useState, useEffect } from 'react';
import { Btn } from './Btn';
import { useResize } from '../../../../hooks';
import { getCountSlides, getMinBreackpoint, getBreackpointsMap } from '../utils'
import { SliderContextProvider, type SliderContextType } from '../context';
import '../style.css';

export type SliderBreackpointProps = {
    slidesPerView?: number,
    marginBetweenSlides?: number,
}
export interface SliderProps extends HTMLAttributes<HTMLDivElement>, SliderBreackpointProps {
    breackpoints?: Record<number, SliderBreackpointProps>,
}

export const Slider: FC<SliderProps> = ({ children = null, slidesPerView = 1, marginBetweenSlides = 0, breackpoints, ...otherProps }) => {
    const sliderWrapperRef = useRef(null);
    const { width: windowWidth } = useResize();
    const [currentSlide, setCurrentSlide] = useState(1);
    const [slideWidth, setSlideWidth] = useState(0);
    const [countSlides, setCountSlides] = useState(0);
    const [countSlidesPerView, setCountSlidesPerView] = useState(slidesPerView);
    const [sliderGap, setSliderGap] = useState(marginBetweenSlides);
    const [translate, setTranslate] = useState(0);
    const [sliderKeys, setSliderKeys] = useState<symbol[]>([]);
    const context: SliderContextType = {
        width: slideWidth,
        margin: sliderGap,
        sliderKeys,
        addSliderKey: (key: symbol) => {
            sliderKeys.push(key);
            setSliderKeys(sliderKeys);
        }
    };

    const calcSlideWidth = () => {
        const count = getCountSlides(children);
        setCountSlides(count);
        if (count > 0 && sliderWrapperRef && sliderWrapperRef.current) {
            const sliderWidth = (sliderWrapperRef.current as HTMLDivElement).clientWidth
            setSlideWidth((sliderWidth - ((countSlidesPerView - 1) * sliderGap)) / countSlidesPerView);
        }
    };
    useEffect(() => {
        if (breackpoints) {
            const bpMap = getBreackpointsMap(breackpoints, windowWidth);
            if (Object.values(bpMap).every((item) => !item)) {
                setCountSlidesPerView(slidesPerView);
                setSliderGap(marginBetweenSlides);
            } else {
                const bpMin = getMinBreackpoint(bpMap);
                const bpTemp = breackpoints[bpMin];
                if (bpTemp.slidesPerView !== undefined) setCountSlidesPerView(bpTemp.slidesPerView);
                if (bpTemp.marginBetweenSlides !== undefined) setSliderGap(bpTemp.marginBetweenSlides);
            }
        }
        calcSlideWidth();
    }, [windowWidth]);
    useEffect(() => {
        calcSlideWidth();
    }, [children, countSlidesPerView, sliderGap]);

    const clickToPrevBtn = () => {
        if (currentSlide > 1) {
            setCurrentSlide(currentSlide - 1)
            setTranslate(translate + (slideWidth + sliderGap));
        }
    };
    const clickToNextBtn = () => {
        if (currentSlide <= countSlides - countSlidesPerView) {
            setCurrentSlide(currentSlide + 1)
            setTranslate(translate - (slideWidth + sliderGap));
        }
    };

    return (
        <SliderContextProvider value={ context }>
            <div className="slider" { ...otherProps }>
                <div className="slider__wrapper" ref={ sliderWrapperRef }>
                    <div
                        className="slider__container"
                        style={ { transform: `translateX(${translate}px)` } }
                    >
                        { children }
                    </div>
                </div>
                <Btn
                    onClick={ clickToPrevBtn }
                    direction="prev"
                    disabled={ currentSlide <= 1 }
                />
                <Btn
                    onClick={ clickToNextBtn }
                    direction="next"
                    disabled={ currentSlide > countSlides - countSlidesPerView }
                />
            </div>
        </SliderContextProvider>
    );
};
