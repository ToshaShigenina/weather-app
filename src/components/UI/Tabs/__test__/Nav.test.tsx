import React, { FC, useState } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { mountTest } from '../../../../tests';
import { Nav, type NavProps } from '../components/Nav';

const testId = 'nav';

describe('Nav', () => {
	mountTest<NavProps>(Nav);

	it('компонент рендерится с тремя кнопками, активна первая', () => {
		const labels = ['Item 1', 'Item 2', 'Item 3'];
		render(<Nav items={ labels } data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		const items = screen.getByTestId(testId).querySelectorAll('li');
		items.forEach((li, i) => {
			if (i === 0) expect(li).toHaveClass('active');
			else expect(li).not.toHaveClass('active');
			expect(li).toHaveClass('tabs__nav-item');
			expect(li.textContent).toContain(labels[i]);
		});
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('компонент рендерится с тремя кнопками, активна вторая', () => {
		const labels = ['Item 1', 'Item 2', 'Item 3'];
		render(<Nav items={ labels } active={ 1 } data-testid={ testId } />);
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		const items = screen.getByTestId(testId).querySelectorAll('li');
		items.forEach((li, i) => {
			if (i === 1) expect(li).toHaveClass('active');
			else expect(li).not.toHaveClass('active');
			expect(li).toHaveClass('tabs__nav-item');
			expect(li.textContent).toContain(labels[i]);
		});
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});

	it('активная кнопка изменяется при клике', () => {
		const labels = ['Item 1', 'Item 2'];
		const onClick = jest.fn();
		const TestComponent: FC = () => {
			const [active, setActive] = useState(0);
			const changeActive = (index: number) => {
				onClick();
				setActive(index);
			};
			return (<Nav
				items={ labels }
				active={ active }
				onClick={ changeActive }
				data-testid={ testId }
			/>);
		};
		render(<TestComponent />);
		expect(screen.getByTestId(testId)).toBeInTheDocument();
		expect(screen.getByTestId(testId)).toMatchSnapshot();
		const items = screen.getByTestId(testId).querySelectorAll('li');
		expect(items[0]).toHaveClass('active');
		expect(items[1]).not.toHaveClass('active');
		const btn = items[1].querySelector('button');
		expect(btn).not.toEqual(null);
		btn && fireEvent.click(btn);
		expect(onClick).toHaveBeenCalled();
		expect(items[0]).not.toHaveClass('active');
		expect(items[1]).toHaveClass('active');
		expect(screen.getByTestId(testId)).toMatchSnapshot();
	});
});