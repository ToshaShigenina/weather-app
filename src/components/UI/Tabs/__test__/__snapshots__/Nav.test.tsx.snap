// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav активная кнопка изменяется при клике 1`] = `
<ul
  class="tabs__nav"
  data-testid="nav"
>
  <li
    class="tabs__nav-item active"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 1
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 2
    </button>
  </li>
</ul>
`;

exports[`Nav активная кнопка изменяется при клике 2`] = `
<ul
  class="tabs__nav"
  data-testid="nav"
>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 1
    </button>
  </li>
  <li
    class="tabs__nav-item active"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 2
    </button>
  </li>
</ul>
`;

exports[`Nav компонент рендерится с тремя кнопками, активна вторая 1`] = `
<ul
  class="tabs__nav"
  data-testid="nav"
>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 1
    </button>
  </li>
  <li
    class="tabs__nav-item active"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 2
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 3
    </button>
  </li>
</ul>
`;

exports[`Nav компонент рендерится с тремя кнопками, активна первая 1`] = `
<ul
  class="tabs__nav"
  data-testid="nav"
>
  <li
    class="tabs__nav-item active"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 1
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 2
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 3
    </button>
  </li>
</ul>
`;

exports[`Nav компонента рендерится с тремя кнопками, активна первая 1`] = `
<ul
  class="tabs__nav"
  data-testid="nav"
>
  <li
    class="tabs__nav-item active"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 1
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 2
    </button>
  </li>
  <li
    class="tabs__nav-item"
  >
    <button
      class="tabs__nav-btn"
    >
      Item 3
    </button>
  </li>
</ul>
`;
